File: src/main/java/ru/belyiz/CodeLinesCounterMojo.java
package ru.belyiz;

import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * Goal to calcutate some coefficients by parsins files in sources directory
 *
 * @goal codeLineCount
 * @phase compile
 */
public class CodeLinesCounterMojo extends AbstractMojo {

    /**
     * Results of calculation within the extensions of processed files
     */
    private Map<String, Counter> fileTypes = new HashMap<String, Counter>();

    /**
     * Path to root folder with source files specified in the POM-file
     * This folder will be used for recursively searching of files and calculate coefficients
     *
     * @parameter expression="${project.build.sourceDirectory}"
     * @readonly
     * @required
     */
    private final File sourceDirectory = new File("");

    /**
     * Encoding of sourse files
     *
     * @parameter expression="${encoding}"
     * default-value="${project.build.sourceEncoding}"
     */
    private final String encoding = "UTF-8";

    @Override
    public void execute() throws MojoExecutionException {
        if (!sourceDirectory.exists()) {
            getLog().error("Source directory \"" + sourceDirectory + "\" is not valid.");
            return;
        }

        FileUtils.processFiles(sourceDirectory, file -> {
            String extension = FileUtils.getFileExtension(file);
            Counter counter = fileTypes.computeIfAbsent(extension, Counter::new);
            String fileAbsolutePath = file.getAbsolutePath();
            try {
                FileUtils.processFileLines(fileAbsolutePath, encoding, counter::processLine);
            } catch (IOException e) {
                getLog().error("Error while reading source file [" + fileAbsolutePath + "]", e);
            }
        });

        writeResultsToLog();
    }

    /**
     * Write calculation results to the process log
     * If count of processed files extensions more than one, then will be printed general result for all processed files
     */
    private void writeResultsToLog() {
        Counter generalCounter = new Counter(".*");

        for (Map.Entry<String, Counter> entry : fileTypes.entrySet()) {
            Counter counter = entry.getValue();
            getLog().info(counter.toString());
            generalCounter.plus(counter);
        }
        if (fileTypes.size() > 1) {
            getLog().info(generalCounter.toString());
        }
    }
}

File: src/main/kotlin/ru/belyiz/CodeLinesCounterMojo.kt
package ru.belyiz

import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugin.MojoExecutionException
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import java.io.File
import java.io.IOException
import java.util.*

/**
 * Goal to calcutate some coefficients by parsins files in sources directory
 */
@Mojo(name = "codeLineCount", defaultPhase = LifecyclePhase.COMPILE)
class CodeLinesCounterMojo : AbstractMojo() {

    /**
     * Results of calculation within the extensions of processed files
     */
    private val fileTypes = HashMap<String, Counter>()

    /**
     * Path to root folder with source files specified in the POM-file
     * This folder will be used for recursively searching of files and calculate coefficients
     */
    @Parameter(defaultValue = "${'$'}{project.build.sourceDirectory}", readonly = true, required = true)
    private val sourceDirectory = File("")

    /**
     * Encoding of sourse files
     */
    @Parameter(defaultValue = "${'$'}{project.build.sourceEncoding}")
    private val encoding = "UTF-8"

    @Throws(MojoExecutionException::class)
    override fun execute() {
        if (!sourceDirectory.exists()) {
            log.error("Source directory \"$sourceDirectory\" is not valid.")
            return
        }

        FileUtils.processFiles(sourceDirectory) { file ->
            val extension = FileUtils.getFileExtension(file)
            val counter = fileTypes.computeIfAbsent(extension, { Counter(it) })
            val fileAbsolutePath = file.absolutePath
            try {
                FileUtils.processFileLines(fileAbsolutePath, encoding, { counter.processLine(it) })
            } catch (e: IOException) {
                log.error("Error while reading source file [$fileAbsolutePath]", e)
            }
        }

        writeResultsToLog()
    }

    /**
     * Write calculation results to the process log
     * If count of processed files extensions more than one, then will be printed general result for all processed files
     */
    private fun writeResultsToLog() {
        var generalCounter = Counter(".*")

        for ((key, counter) in fileTypes) {
            log.info(counter.toString())
            generalCounter += counter
        }
        if (fileTypes.size > 1) {
            log.info(generalCounter.toString())
        }
    }
}